Lib/typing.py

```diff
@@ -823,6 +874,117 @@ def __init__(self, name, *constraints, bound=None,
             self.__module__ = def_mod
 
 
+class TypeVarTuple(_Final, _Immutable, _root=True):
+    """Type variable tuple.
+
+    Usage:
+
+      Ts = TypeVarTuple('Ts')  # Can be given any name
+
+    Just as a TypeVar (type variable) is a placeholder for a single type,
+    a TypeVarTuple is a placeholder for an *arbitrary* number of types. For
+    example, if we define a generic class using a TypeVarTuple:
+
+      class C(Generic[*Ts]): ...
+
+    Then we can parameterize that class with an arbitrary number of type
+    arguments:
+
+      C[int]       # Fine
+      C[int, str]  # Also fine
+      C[()]        # Even this is fine
+
+    For more details, see PEP 646.
+    """
+
+    def __init__(self, name):
+        self._name = name
+        self._unpacked = UnpackedTypeVarTuple(name, self)
+
+    def __iter__(self):
+        yield self._unpacked
+
+    def __repr__(self):
+        return self._name
+
+
+class UnpackedTypeVarTuple(_Final, _Immutable, _root=True):
+
+    def __init__(self, name, packed):
+        self._name = name
+        self._packed = packed
+        # __parameters__ will be use to figure out what the __parameters__
+        # of e.g. `tuple[*Ts]` should be. In the case of an unpacked
+        # TypeVarTuple, it's unclear whether this should refer to the
+        # UnpackedTypeVarTuple or the TypeVarTuple itself. We somewhat
+        # arbitrarily decide to go for the latter.
+        self.__parameters__ = (packed,)
+
+    def __repr__(self):
+        return '*' + self._name
+
+
+class StarredTuple(_Final, _Immutable, _root=True):
+    """Implementation of starred tuple for older versions of Python.
+
+    A starred tuple is e.g. tuple[*tuple[int, str]]. In newer versions of
+    Python, this is handled through a change to `tuple` itself.
+    """
+
+    def __init__(self, tup):
+        self.__origin__ = tuple
+        self._tuple = tup
+        self.__args__ = tup.__args__
+        self.__parameters__ = tup.__parameters__
+
+    def __repr__(self):
+        return '*' + repr(self._tuple)
+
+    def __eq__(self, other):
+        if isinstance(other, StarredTuple):
+            return self._tuple == other._tuple
+        else:
+            return False
+
+
+class Unpack(_Final, _Immutable, _root=True):
```

Comment:
Could this be implemented with the `_SpecialForm` decorator, like Final and ClassVar?

Reply:
Oh, neat! I didn't know about `_SpecialForm`. It does indeed feel much more consistent with the rest of `typing.py` to do it this way, and even simplifies a few checks. Great call! :)